src/GeniusYield/TxBuilder/IO/Query.hs

```diff
@@ -0,0 +1,152 @@
+{-|
+Module      : GeniusYield.TxBuilder.IO.Query
+Copyright   : (c) 2023 GYELD GMBH
+License     : Apache 2.0
+Maintainer  : support@geniusyield.co
+Stability   : develop
+
+**INTERNAL MODULE**
+-}
+module GeniusYield.TxBuilder.IO.Query (
+    GYTxQueryMonadIO,
+    runGYTxQueryMonadIO,
+    ioToQueryMonad
+) where
+
+import           Control.Monad.Reader
+import           GHC.Stack                    (withFrozenCallStack)
+
+import           GeniusYield.Imports
+import           GeniusYield.TxBuilder.Class
+import           GeniusYield.TxBuilder.Errors
+import           GeniusYield.Types
+
+
+-------------------------------------------------------------------------------
+-- GY implementation
+-------------------------------------------------------------------------------
+
+-- | 'GYTxQueryMonad' interpretation run under IO.
+type role GYTxQueryMonadIO representational
+newtype GYTxQueryMonadIO a = GYTxQueryMonadIO { runGYTxQueryMonadIO' :: GYTxQueryIOEnv -> IO a }
```

Comment:
Why can't it be a `newtype` wrapper around `ReaderT GYTxQueryIOEnv IO a`?

Reply:
But even o/w let's not bother about making this change now as we already used similar representation earlier.

Reply:
This is what I started with indeed! But actually it changes the type roles. The `a` in `ReaderT` is set to `nominal` (we cannot change this). We're looking for representational (to allow coercions). The function arrow (`(->)`) doesn't have this restriction.